[5.2] Add redirect helper to router

brayniverse opened the pull request:
As proposed on [this internals thread](https://github.com/laravel/internals/issues/172), this PR adds a redirect helper to the Router for those occasions when you want to redirect from an old route to a new route.

brayniverse committed:
Add redirect helper

As proposed on this internals thread, https://github.com/laravel/internals/issues/172.
This commit adds a redirect helper to the Router for those occasions where you want
to redirect from an old route to a new route.

brayniverse committed:
Remove whitespace

crynobone commented on src/Illuminate/Routing/Router.php:
```diff
@@ -360,6 +361,21 @@ public function resource($name, $controller, array $options = [])
+    {
+        return $this->get($from, function () use ($to, $status) {
+            return new RedirectResponse($to, $status);
+        });
```
Wouldn't this make it impossible to use `php artisan route:cache` since the command can't compile closure based routes?

brayniverse replied:
That is true, same as if you wanted to redirect using a closure though. The only way around it is to handle your own redirects in a controller.

Garbee replied:
Difference is, the cacher should be able to handle this by-default. So a method should be found to allow for redirects to be cached by default. Perhaps an internal controller that has the redirect method and signature?

Redirect should not come with the warning that in order to cache you handle it manually, that makes it not nearly as useful as it would otherwise be.

brayniverse replied:
Internal redirect controller sounds like a good solution. Thanks @Garbee, I'll implement it later tonight.

JosephSilber replied:
@Garbee how would you tell the internal controller which URL to redirect to?

brayniverse replied:
@JosephSilber you can't, I'm trying to figure out a different solution. Any suggestions are welcome.

brayniverse replied:
It's a bit of a stretch, but I have a solution.

Controllers can specify an optional `callAction` method which will be called _if present_. The method is given the `$method` and `$parameters` as it's arguments, additionally, we could provide the `Illuminate\Routing\Route` instance when calling it in `Illuminate\Routing\ControllerDispatcher::dispatch`. By doing so, we would have access to the routes `getAction()` method, which allows us to access `data` from the following:

```php
public function redirect($uri, $destination, $status = 301)
{
    return $this->any($uri, [
        'uses' => '\Illuminate\Foundation\Http\RedirectController@handle`,
        'data'' => compact('destination', 'status'),
    ]);
}
```

The `callAction` method would look like such:

```php
public function callAction($method, $parameters, $route)
{
    return $this->handle(
        $route->getAction()['data']['destination'],
        $route->getAction()['data']['status']
    );
}
```

tomschlick commented on src/Illuminate/Routing/Router.php:
```diff
@@ -360,6 +361,21 @@ public function resource($name, $controller, array $options = [])
+     */
+    public function redirect($from, $to, $status = 301)
+    {
+        return $this->get($from, function () use ($to, $status) {
```
This should probably be a `$this->any()` call as you could be doing a 303/307 redirect from a POST/PUT.

brayniverse committed:
Use RedirectController for handling redirect

Using a closure for handling the redirect prevents you from caching the routes.
This commit uses an internal controller to handle the redirect so we can cache routes.

brayniverse committed:
allow setting and getting additional data

This allows us to get and set additional data which will be available on the route instance.

brayniverse committed:
remove whitespace

brayniverse committed:
fix typo

brayniverse committed:
remove unused use statement

brayniverse committed:
Create RedirectController.php

GrahamCampbell commented:
Interesting idea. Perhapse 5.3 is a better target though.

brayniverse commented:
Sure thing. I thought it wouldn't be a breaking change so I could backport it to 5.2. I'll close this and recreate for 5.3.

GrahamCampbell commented:
Larave 5.3 is essentially the current stable at this point.
